casper/src/main/scala/coop/rchain/casper/util/comm/CommUtil.scala

```diff
@@ -52,41 +56,117 @@ object CommUtil {
       _     <- TransportLayer[F].broadcast(peers, msg)
     } yield ()
 
+  def requestApprovedBlock[
+      F[_]: Monad: Capture: MultiParentCasperConstructor: Log: Time: Metrics: TransportLayer: NodeDiscovery: ErrorHandler: PacketHandler]
+    : F[Unit] = {
+    val request = ApprovedBlockRequest("PleaseSendMeAnApprovedBlock").toByteString
+
+    def askPeers(peers: List[PeerNode], local: PeerNode): F[Unit] = peers match {
+      case peer :: rest =>
+        for {
+          _    <- Log[F].info(s"Sending request for ApprovedBlock to $peer")
+          send <- TransportLayer[F].roundTrip(peer, packet(local, request), 5.seconds)
+          _ <- send match {
+                case Left(err) =>
+                  Log[F].info(s"Failed to get response from $peer because: $err") *>
+                    askPeers(rest, local)
+
+                case Right(response) =>
+                  Log[F].info(s"Received response from $peer! Dispatching...")
+                  Connect
+                    .dispatch[F](response)
+                    .flatMap(_ => {
+                      MultiParentCasperConstructor[F].lastApprovedBlock.map {
+                        case Some(_) => () //got one!
+                        case None    => askPeers(rest, local)
+                      }
+                    })
```

Comment:
@sebtomba So with this we are introducing a synchronous two way communication at the Casper layer.

Reply:
@KentShikama @birchmd @rabbitonweb I believe this is a bad idea. Let us discuss it because this is important.

Reply:
I would like to remove the `roundTrip` from the transport layer and use synchronous communication only for Kademlia actions. However, this requires that Casper reacts only to events. The function `askPeers` is pretty heavy. It does a `roundTrip` dispatches the response and calls `askPeers` recursively. This function is then called in `NodeRuntime` every few seconds. There must exist smarter way to propagate approved blocks.

Reply:
@sebtomba The function is only expensive in the case that the node doesn't already have an approved block. If it had an approved block then none of that communication is done. That being said, repeating the check to see if we have an approved block over and over again is unnecessary and I would rather not do it. I am also completely on board with having Casper being entirely event-based. This initialization step where we need to get an approved block to start Casper is a bit of a special case though because we cannot react to any other event until we have an approved block.

Reply:
Having an in-depth discussion about how comm is being re-designed and how Casper should use it most effectively is welcome. Perhaps we should schedule a time to talk about it for this week.